treq/auth.py

```diff
@@ -26,6 +182,112 @@ def request(self, method, uri, headers=None, bodyProducer=None):
             method, uri, headers=headers, bodyProducer=bodyProducer)
 
 
+class _RequestDigestAuthenticationAgent(object):
+
+    digest_auth_cache = {}
```

Comment:
It looks like requests also put the shared state on the HTTPDigestAuth object (https://github.com/kennethreitz/requests/blob/master/requests/auth.py#L79).

Probably start with a couple of test cases that exercise the desired sharing behavior, and then start moving the digest_auth_cache and to the HTTPDigestAuth instance. Probably reasonable to put the header construction code there too. And just have the `_RequestDigestAuthenticationAgent` ask the credentials for the header value.